components/Builder.tsx

```diff
@@ -169,38 +176,57 @@ export default function Builder({ data }: { data: any }) {
         />
       </div>
 
-      <div className="flex gap-2 justify-center items-center">
-        {isSignedIn && (
-          <SaveButton
-            formation={formation}
-            artifact={artifact}
-            layout={layout}
-            user={user}
+      <div className="flex flex-col gap-2 justify-center items-center w-full">
+        <div className="flex gap-2">
+          {isSignedIn && (
+            <SaveButton
+              artifact={artifact}
+              formation={formation}
+              layout={layout}
+              user={user}
+            />
+          )}
+          <Button onClick={onDownloadButtonClick} className="h-8 px-2">
+            <Download />
+          </Button>
+        </div>
+        <div className="flex gap-2 justify-between w-full pr-4">
+          <Input
+            className="w-28"
+            hasClearInput
+            onChange={(e) =>
+              setCharacterFilter((prev) => ({ ...prev, name: e.target.value }))
+            }
+            onClearClick={() => {
+              setCharacterFilter((prev) => ({ ...prev, name: "" }));
+              searchInputRef.current?.focus();
+            }}
+            placeholder="Search..."
+            ref={searchInputRef}
+            value={characterFilter.name}
           />
-        )}
-        <Button onClick={onDownloadButtonClick} className="h-8 px-2">
-          <Download />
-        </Button>
-        <CharacterFilter
-          classes={data.classes}
-          factions={data.factions}
-          characterFilter={characterFilter}
-          updateCharacterFilter={setCharacterFilter}
-          className="relative -ml-12 left-[4.5rem] md:left-[10.5rem]"
-        />
+          <CharacterFilter
+            characterFilter={characterFilter}
+            classes={data.classes}
+            factions={data.factions}
+            updateCharacterFilter={setCharacterFilter}
+          />
+        </div>
       </div>
 
       <ScrollArea
-        className="flex flex-col items-center"
+        className="w-full mt-2"
         style={{
           height: `calc(100vh - ${layoutHeights[layout] ?? "24"}rem - 5rem)`,
         }}
       >
-        <div className={`grid grid-cols-5 sm:grid-cols-10 gap-2 pt-2 mx-6`}>
+        <div className="grid grid-cols-[repeat(auto-fit,minmax(56px,1fr))] gap-2 pr-4">
```

Comment:
i'm surprised https://tailwindcss.com/docs/grid-auto-columns this didn't work here

Reply:
It sounds like it would but this actually just defines the width of undefined columns, not sets an automatic number of columns